Add support for extra fields

d-miketa opened the pull request:
Jira boards are very customisable. This PR adds the ability to declare extra UDA fields from `bugwarriorrc`.

https://github.com/ralphbean/bugwarrior/issues/860

d-miketa committed:
gitignore

d-miketa committed:
feat: jira extra field support

ryneeverett commented on bugwarrior/services/jira.py:
```diff
@@ -174,10 +237,21 @@ def to_taskwarrior(self):
 
+        extrafields = self.get_extrafields()
+
+        return {**fixed_fields, **extrafields}
```
To me this looks weird -- can we please consistently include or omit an underscore in these variables?

Personally I'd just add `**self.get_extrafields(),` to the end of the existing dict, eliminating the need for variables. But I leave that to your discretion.

d-miketa replied:
Yeah I agree and have fixed it.

The reason for this format was because I originally wanted to implement this for fields of the form `customfield_12345`, so I called the variable `customfields`. Then I realised I can target _any_ field, so switched to the more generic `extrafields`. But of course now that no longer makes sense!

ryneeverett commented on bugwarrior/services/jira.py:
```diff
@@ -16,6 +17,66 @@
+
+    @property
+    def id(self):
+        return self.keys[0]
```
Shouldn't this be `'.'.join(keys)` since we don't know which key caused the `KeyError`? (I'm assuming that's the only usage of this method.)

ryneeverett commented on bugwarrior/services/jira.py:
```diff
@@ -16,6 +17,66 @@
+        return extrafields
+
+
+@pydantic.dataclasses.dataclass
```
Couldn't we use a standard library `dataclasses.dataclass` here? I guess we still support python3.6, which doesn't include dataclasses, so maybe just add a comment that we can switch once we drop 3.6?

ryneeverett reviewed (approved):
I'd like to see my comments addressed before merging but they're fairly trivial in the scope of this patch.

d-miketa committed:
resolve reviewer's comments

extrafields -> extra_fields
change JiraExtraField key retrieval error
comment about pydantic dataclass (remove after 3.6)

d-miketa committed:
resolve reviewer's comments

extrafields -> extra_fields
change JiraExtraField key retrieval error
comment about pydantic dataclass (remove after 3.6)

d-miketa committed:
inline variable

d-miketa committed:
extrafield -> extra_field

accidentally nuked this change from earlier commit

ryneeverett commented:
Merged in 55db4c4a212896f5807ea61ab95ace434296f080.

ryneeverett commented on bugwarrior/services/jira.py:
```diff
@@ -33,7 +94,10 @@ class JiraConfig(config.ServiceConfig, prefix='jira'):
 
-    @pydantic.root_validator
+    extra_fields: JiraExtraFields = pydantic.Field(
+        default_factory=lambda: JiraExtraFields(''))
```
Note that in the rebase I changed this to `extra_fields: typing.Optional[JiraExtraFields] = None` for consistency with other non-required fields with empty defaults.
